tsl/src/chunk_copy.c

```diff
@@ -0,0 +1,581 @@
+
+	/* It has to be a foreign table chunk */
+	if (cc->chunk->relkind != RELKIND_FOREIGN_TABLE)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("\"%s\" is not a valid remote chunk", get_rel_name(chunk_relid))));
+
+	/* It has to be an uncompressed chunk, we query the status field on the AN for this */
+	if (ts_chunk_is_compressed(cc->chunk))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("\"%s\" is a compressed remote chunk. Chunk copy/move not supported"
+						" currently on compressed chunks",
+						get_rel_name(chunk_relid))));
+
+	ht = ts_hypertable_cache_get_cache_and_entry(cc->chunk->hypertable_relid,
+												 CACHE_FLAG_NONE,
+												 &hcache);
+
+	ts_hypertable_permissions_check(ht->main_table_relid, GetUserId());
+
+	if (!hypertable_is_distributed(ht))
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("hypertable \"%s\" is not distributed",
+						get_rel_name(ht->main_table_relid))));
+
+	/* Check that src_node is a valid DN and that chunk exists on it */
+	cc->src_server = data_node_get_foreign_server(src_node, ACL_USAGE, true, false);
+	Assert(NULL != cc->src_server);
+
+	if (!ts_chunk_has_data_node(cc->chunk, src_node))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("chunk \"%s\" does not exist on source data node \"%s\"",
+						get_rel_name(chunk_relid),
+						src_node)));
+
+	/* Check that dst_node is a valid DN and that chunk does not exist on it */
+	cc->dst_server = data_node_get_foreign_server(dst_node, ACL_USAGE, true, false);
+	Assert(NULL != cc->dst_server);
+
+	if (ts_chunk_has_data_node(cc->chunk, dst_node))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("chunk \"%s\" already exists on destination data node \"%s\"",
+						get_rel_name(chunk_relid),
+						dst_node)));
+
+	/*
+	 * Populate the ChunkCopy structure for use by various stages
+	 *
+	 * Get the operation id for this chunk move/copy activity. The naming
+	 * convention is "ts_copy_seq-id_chunk-id_table-name_schema-name and it can
+	 * get truncated due to NAMEDATALEN restrictions
+	 */
+	cc->id = ts_catalog_table_next_seq_id(ts_catalog_get(), CHUNK_COPY_ACTIVITY);
+	snprintf(cc->operation_id.data,
+			 sizeof(cc->operation_id.data),
+			 "ts_copy_%d_%d_%s_%s",
+			 cc->id,
+			 cc->chunk->fd.id,
+			 NameStr(cc->chunk->fd.table_name),
+			 NameStr(cc->chunk->fd.schema_name));
+	cc->src_node = pstrdup(src_node);
+	cc->dst_node = pstrdup(dst_node);
+	cc->delete_on_src_node = delete_on_src_node;
+	cc->stage = NULL;
+
+	/* Persist the entry in the catalog */
+	chunk_copy_activity_insert(cc);
+
+	ts_cache_release(hcache);
+	MemoryContextSwitchTo(old);
+
+	/* Commit to get out of starting transaction */
+	PopActiveSnapshot();
+	CommitTransactionCommand();
+}
+
+static void
+chunk_copy_cleanup(ChunkCopy *cc)
+{
+	/* Done using this long lived memory context */
+	MemoryContextReset(cc->mcxt);
```

Comment:
Since you allocated a new context, should this be:

```suggestion
	MemoryContextDelete(cc->mcxt);
```